### PR TITLE
Add collapsible navbar functionality

### PR DESCRIPTION
**Pull request by bradh11**
# Closes 
* Issue 5106: https://github.com/nautobot/nautobot/issues/5106
* Issue 5183 https://github.com/nautobot/nautobot/issues/5183
* Issue 5285 https://github.com/nautobot/nautobot/issues/5285
* Issue 5308 https://github.com/nautobot/nautobot/issues/5308
* Discussion: https://github.com/orgs/nautobot/discussions/3872#discussioncomment-8186580
# What's Changed
This PR brings forth a collapse capable navbar behavior. Users will often want to have more screen real estate for their work or want to be able to leverage a responsive mobile view. 

**This PR brings forth the following capabilities:**
Added:
* **Collapse Capable Side Navbar:** Side Navbar is now able to be expanded and collapsed
* **Expandable Main Content:** The Main Content part of the UI grows as the Side Navbar collapses and shrinks as the Side Navbar expands.
* **Better mobile friendly bottom footer:** Bottom footer currently does not handle responsive media queries properly and will overlap as the browser width gets too small. This update will switch to vertically aligned bottom nav menu items once a certain media query is hit.
* **Automatic Collapsing Navbar for mobile:** The Side Navbar will auto-collapse at a certain media width which is a better experience for mobile devices.

Fixes:
* **Brand Icon mouseover Background:** Fix for mouseover effect on the Brand / Icon (was flashing white background vs being transparent) when in dark mode.

# Screenshots
https://github.com/nautobot/nautobot/assets/18483910/2528a394-863a-45c4-83e6-3b6e81e4f875

https://github.com/nautobot/nautobot/assets/18483910/cec88e23-b2e4-4fbc-84bc-d6f2ff770d02

https://github.com/nautobot/nautobot/assets/18483910/cfbe4228-113a-4dad-a085-2d612d9e1fb6

# TODO
There are several items which are still outstanding in this PR. 

- [X] Need to merge styles to `base.css` and `dark.css`
- [X] Ensure responsive breakpoints are properly configured for the bottom navbar.

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Comment by glennmatthews** [comment]
This also addresses #5106, it sounds like :-)

**Comment by lampwins** [comment]
How does this look on mobile?

**Comment by bradh11** [comment]
> How does this look on mobile?

It looks good.  In order to make it better I should probably make the bottom nav more responsive as well.  I noticed when it gets tight, the content starts to overlap.  The responsive view for bottom nav on mobile should switch to vertical list view rather than overflow.

Here is what it looks like on iPhone 14 Pro
image

**Comment by bradh11** [comment]
I'm seeing a CI failure for tests for the views, but I am not sure where I need to go to adjust the test. Appreciate some input on how to adjust the test.  Thanks!

**Comment by bradh11** [comment]
Ok CI testing issue is resolved.  This is ready for review.

**Comment by matt852** [comment]
To add this this, I opened issue #5308 today.  When logging into Nautobot on a mobile browser, the interface is pretty much unusable as the navbar takes up the entire screen.

This may need to be addressed in a separate PR, but can we maybe update the navbar to automatically scale, or collapse on mobile?

**Comment by bradh11** [comment]
> To add this this, I opened issue #5308 today. When logging into Nautobot on a mobile browser, the interface is pretty much unusable as the navbar takes up the entire screen.
> 
> This may need to be addressed in a separate PR, but can we maybe update the navbar to automatically scale, or collapse on mobile?

This PR handles mobile views quite nicely however it is not automatically detecting the viewport and hiding the nav. The toggle button to hide it is able to be leveraged to toggle it in short term, but if we agree it makes sense to automatically hide it once you hit a certain viewport size, it can fairly easily be accomplished.  I will defer to the Nautobot Core team on whether they want to see this within this PR, or a separate once.  @glennmatthews - any preference?

**Comment by bradh11** [comment]
@matt852 in the meantime, try cloning this branch and test it out to see the mobile experience.  Would love your input.

File: nautobot/project-static/css/base.css
Comment (glennmatthews) [inline nautobot/project-static/css/base.css:930]: Are these correct? It feels weird to me that we'd add a `.collapsed` class to either of these areas.

File: nautobot/project-static/css/base.css
Comment (glennmatthews) [inline nautobot/project-static/css/base.css:799]: Can we clean up and put these in appropriate order with the existing contents of the file? Don't want the organizational work from #5268 to go to waste. :-)

File: nautobot/core/templates/search_form.html
Comment (glennmatthews) [inline nautobot/core/templates/search_form.html:2]: This seems to have broken the layout of the search form on the home page on mobile.

Before:

image

After:

image

File: nautobot/core/templates/inc/nav_menu.html
Comment (bradh11) [reply nautobot/core/templates/inc/nav_menu.html:18]: yes, I will make that adjustment.

File: nautobot/project-static/css/base.css
Comment (bradh11) [reply nautobot/project-static/css/base.css:887]: This is the styling coming from Nautobot Cloud Console - I manually moved the styling over.  That said, it was part of the CSS framework on NB Cloud Console, so the complexities were hidden. I will investigate the usage of using some of the pre-existing button styling vs creating a special case here.

File: nautobot/project-static/css/base.css
Comment (bradh11) [reply nautobot/project-static/css/base.css:799]: will do!

File: changes/5172.fixed
Comment (bradh11) [reply changes/5172.fixed:2]: #5268 accomplishes button wrapping avoidance by adjusting gap and margins.  This PR was going to ensure there is no `text-wrap` assigned to navbar button menu items, but in further review I think it may no longer need to be added.

File: nautobot/core/templates/inc/nav_menu.html
Comment (bradh11) [reply nautobot/core/templates/inc/nav_menu.html:135]: i think it makes sense to do this.  exception being when clicking the brand icon

File: nautobot/core/templates/search_form.html
Comment (bradh11) [reply nautobot/core/templates/search_form.html:2]: this may be resolved with the auto collapse breakpoint on window size < 1007px

File: nautobot/core/templates/inc/nav_menu.html
Comment (bradh11) [reply nautobot/core/templates/inc/nav_menu.html]: i set this to -60px and it is working as desired now

File: nautobot/project-static/css/base.css
Comment (bradh11) [reply nautobot/project-static/css/base.css:930]: @glennmatthews - these are required so that the banner and breadcrumbs do not overlap the navbar-header and navbar-toggler button.  they keep the two elements in the same position even when the navbar is toggled to the collapsed state.

File: nautobot/core/templates/inc/nav_menu.html
Comment (bradh11) [reply nautobot/core/templates/inc/nav_menu.html:135]: addressed in commit https://github.com/nautobot/nautobot/pull/5172/commits/0937825362054b0ef7e2097f750505fc4c07ec3f

File: nautobot/core/templates/search_form.html
Comment (bradh11) [reply nautobot/core/templates/search_form.html:2]: confirmed that this is no longer an issue when navbar auto collapses.  Please have a look and let me know if the new behavior is acceptable for addressing this.

**Comment by glennmatthews** [comment]
Revamped some more. Updated screenshots:

image

image

image

image

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**
